polkadot/node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -2107,6 +2119,101 @@ async fn handle_collation_fetch_response(
 	result
 }
 
+// Returns the claim queue without fetched or pending advertisement. The resulting `Vec` keeps the
+// order in the claim queue so the earlier an element is located in the `Vec` the higher its
+// priority is.
+fn unfulfilled_claim_queue_entries(relay_parent: &Hash, state: &State) -> Result<Vec<ParaId>> {
+	let relay_parent_state = state
+		.per_relay_parent
+		.get(relay_parent)
+		.ok_or(Error::RelayParentStateNotFound)?;
+	let scheduled_paras = relay_parent_state.assignment.current.iter().collect::<HashSet<_>>();
+	let paths = state.implicit_view.paths_via_relay_parent(relay_parent);
+
+	let mut claim_queue_states = Vec::new();
+	for path in paths {
+		let mut cq_state = ClaimQueueState::new();
+		for ancestor in &path {
+			cq_state.add_leaf(
+				&ancestor,
+				&state
+					.per_relay_parent
+					.get(&ancestor)
+					.ok_or(Error::RelayParentStateNotFound)?
+					.assignment
+					.current,
+			);
+
+			for para_id in &scheduled_paras {
+				let seconded_and_pending = state.seconded_and_pending_for_para(&ancestor, &para_id);
+				for _ in 0..seconded_and_pending {
+					cq_state.claim_at(&ancestor, &para_id);
+				}
+			}
+		}
+		claim_queue_states.push(cq_state);
+	}
+
+	let unfulfilled_entries = claim_queue_states
+		.iter_mut()
+		.map(|cq| cq.unclaimed_at(relay_parent))
+		.max_by(|a, b| a.len().cmp(&b.len()))
```

Comment:
First to clear the problem I am trying to solve here.

`unfulfilled_claim_queue_entries` is used by `get_next_collation_to_fetch` and it's purpose is to produce a `Vec` of `ParaID`s which can be fetched, ordered by urgency. So a result could look like `[A A B]` which means that we can accept collations from both paras A and B but A is strongly preferred in this case.

As you mentioned the claims on different forks can be quite different and from all forks (all claim_queue_states) we need to produce a single `Vec` with `ParaId`s ordered by priority. Hope we are on the same page so far.

Now why `max_by`?
No matter how many forks we have got, they should have the same claims because the claim queue should be the same for them. Let's say we have got two forks - X1 and X2. They should have the same claim queue state (`[A A B]`). The difference might be that at X1 first spot might already be claimed while in X2 it might not be claimed. So in this case the unfulfilled entries for X1 will be `[AB]` while at X2 - `[AAB]`.

For this reason I naively decided to go for the longest unfulfilled entries Vec since it might be the most urgent one to satisfy. Most urgent because if it is the longest one it's first slot is PROBABLY not satisfied. Of course there are a lot of corner cases here. E.g. the second slot might have been claimed at X1 which doesn't make it less urgent than X2.

We can be smarter here and examine what's claimed at each fork and make a more informed decision but I don't think it's worth the complication.

Also it's worth adding this information as a comment in the code because it's not obvious.

Reply:
(added a comment)

Reply:
Isn't that essentially incentivising forks? We will attempt to fetch for the fork which the most unfulfilled claims.

Reply:
Changed the code to accept an advertisement if it has got a free claim at at least one fork.